### ISSUE TITLE
[HOLD for payment 2022-01-11] [$1000] Copying messages with markdown via CTRL + C does not keep formatting

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#5142]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Send several messages each with a different markdown (bold, strike, italic, and hyperlink)
2. Copy and paste each of the messages in compose box via the ctrl +C method 

## Expected Result:
HTML Markdown should be copied and pasted.

## Actual Result:
HTML Markdown is not copied and user has to add the markdown again.

## Workaround:
HTML Markdown is not copied and user has to add the markdown again.

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.0.94-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

![image](https://user-images.githubusercontent.com/44479856/132577560-fb0384ef-f734-4763-9fb7-89f033b1e1ab.png)

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Jag96 / @parasharrajat 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1631123012073500?thread_ts=1631122717.073400&cid=C01GTK53T8Q

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @nickmurray47 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by isagoico** [comment]
Comment from [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1631123012073500?thread_ts=1631122717.073400&cid=C01GTK53T8Q) that has more information on this bug:  

- From @Jag96 

Note that this is a different issue from https://github.com/Expensify/App/issues/5125, which handles the copy via Copy from clipboard. Since this has come up a few times, I'm filing an issue here so we can have something to reference in the future.
The credit for finding this bug goes to @Rajat Parashar who asked about it initially in [this issue](https://github.com/Expensify/App/pull/4009#issuecomment-879534328).
The issue here is:

> When I copy the existing sent message which is formatted and pastes it. it does not convert to formatted MD strings. The main issue here is that we don't use HTML semantic tags for rendering Every type of element is made with div and styling rules. Which we can't really interpret ATM.

So in order to fix this issue, we would have to use html tags such as a/em/b in our app instead of divs + styling.

**Comment by parasharrajat** [comment]
@isagoico You tagged the wrong guy.

**Comment by nickmurray47** [comment]
Reproduction steps are clear and confirmed this still happens locally.

> So in order to fix this issue, we would have to use html tags such as a/em/b in our app instead of divs + styling.

@parasharrajat Where do these adjustments need to be made and are you interested in proposing a more detailed solution?

**Comment by parasharrajat** [comment]
I have requested the lib owner to possibly upgrade the lib so that we get this support out of the box. Otherwise, I will update here on what I think is best.

**Comment by nickmurray47** [comment]
Great, thanks @parasharrajat!

**Comment by nickmurray47** [comment]
Going to assign the `External` label and unassign myself, seems like a good issue for a contributor

**Comment by MelvinBot** [comment]
Triggered auto assignment to @dylanexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by dylanexpensify** [comment]
Working on this now!

**Comment by dylanexpensify** [comment]
Job has been posted! 🎉 

Internal: https://www.upwork.com/ab/applicants/1437772705400451072/job-details
External: https://www.upwork.com/jobs/~01ba05e404d040122d

**Comment by MelvinBot** [comment]
Triggered auto assignment to @timszot (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by dylanexpensify** [comment]
not overdue

**Comment by dylanexpensify** [comment]
not overdue

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by timszot** [comment]
No proposals here yet.

**Comment by mallenexpensify** [comment]
https://www.upwork.com/jobs/~01396fef12d0e84b11
Created new job (other was auto-closed), raised price to $500

**Comment by monykini** [comment]
This issue isn't related to the copied content , It's <textarea> not allowing html content inside 
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea , 
The HTML <textarea> element represents a multi-line plain-text editing control.
     **_Permitted content Text_**
Now how do you handle this ? 
There are multiple solutions , 
One -> convert the textarea to div with contenteditable="true" and apply the same styles to that and the same events of course  ( the easy way out , but a feasible one )
Two -> On paste , convert the html to a image blob set it as background of the textarea 
Three -> Create a custom html element with the possibility of rendering direct html ( the long way )
There are multiple **_challenges_** to this solution , but only in the case if html tags aren't stripped if present on input for the bot to work , since with these solutions the text can be something like <span class="bold" > text </span>

**Comment by mallenexpensify** [comment]
@timszot can you please review @monykini 's proposal above! 

Also... 
We are placing all non-critical issues on hold while we're on a company offsite.  The hold is expected to be lifted on 11/19 (but could go longer).  For any PRs that are submitted before or during the hold, we will add a $250 bonus payment.

**Comment by timszot** [comment]
@monykini I like your `option one` proposal. Can you go into more detail on where in the code exactly you'd be making changes?

**Comment by kadiealexander** [comment]
Doubled job price again [here](https://www.upwork.com/jobs/~01396fef12d0e84b11) for the team!

**Comment by dylanexpensify** [comment]
not overdue

**Comment by MelvinBot** [comment]
Triggered auto assignment to @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @thienlnam (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by timszot** [comment]
Reassigned this one as I'll be going OOO.

**Comment by kursat** [comment]
# Proposal

I noticed in the DOM of messages with markdown, we have `data-testid` attribute containing tag name of markdown message. So by intercepting copy shortcut, we can parse selection to markdown.

In my proposal, I'll include context menu copy feature, also doesn't copy messages with markdown.

### Steps

1. Create a shortcut for `Ctrl+C` to override default behavior.
2. Get HTML of selection via Selection API.
3. Clear unnecessary attributes.
4. Replace tags with tag name inside `data-testid` attribute.
5. Parse generated clean HTML with `ExpensiMark`.


## 1. Fixing Ctrl+C behaviour

We have 2 cases here:

1. Copying from a single message.
2. Copying from multiple messages.

To be able to copy from multiple messages, we need to use Selection API. At this point please consider, there is a SelectionParser library. I'll explain it at the end, since it does most of the job.

In `ReportActionsView.js` file we register a shortcut at `componentDidMount` function, and remove at `componentWillUnmount`.

```js

const copyShortcutModifiers = ['control'];

this.unsubscribeCopyShortcut = KeyboardShortcut.subscribe('C', () => {
    this.copySelectionToClipboard();
}, copyShortcutModifiers, false);
```

```js
if (this.unsubscribeCopyShortcut) {
    this.unsubscribeCopyShortcut();
}
```
The function that does the job should be like this:

```js
copySelectionToClipboard = () => {
    const selectionMarkdown = SelectionParser.parseToMarkdown();

    Clipboard.setString(selectionMarkdown);
}
```

You can see all changes here:
https://github.com/kursat/App/commit/84f4ba18f9face9bc792999c8a9bbd6253e6db54#diff-3d3d60df30d1761a83ff8413b06ecef63fcd0084053cd70d5686098e71d87362

## Fixing context menu

With copy to clipboard feature on the context menu we have 3 functions:

1. If there is a selection, gets selection as string and sets to clipboard.
2. If there is no selection, gets message HTML and decodes as string and sets to clipboard.
3. If message is attachment copies, gets message HTML, but sets to clipboard as string. 

This is current behavior:

https://user-images.githubusercontent.com/1279379/143318869-df704438-acb2-4302-8d25-9b8b433197da.mp4

IMHO these should work like this:

1. Should parse selection and copy the markdown
2. Parse all message to markdown
3. Should copy only link of attachment (Not present in the proposal code) 

To achieve this we should change following function
 
https://github.com/Expensify/App/blob/cc78af22380831563e71d5914f89e95cdc70bc26/src/pages/home/report/ContextMenu/ContextMenuActions.js#L50-L68

to this:
```js
onPress: (closePopover, {reportAction, selection}) => {
    const message = _.last(lodashGet(reportAction, 'message', null));
    const html = lodashGet(message, 'html', '');

    const parser = new ExpensiMark();
    const reportMarkdown = parser.htmlToMarkdown(html);

    const text = selection || reportMarkdown;

    const isAttachment = _.has(reportAction, 'isAttachment')
        ? reportAction.isAttachment
        : isReportMessageAttachment(text);
    if (!isAttachment) {
        Clipboard.setString(text);
    } else {
        Clipboard.setString(html);
    }
    if (closePopover) {
        hideContextMenu(true, ReportActionComposeFocusManager.focus);
    }
},
```

And we should parse selection as markdown at `PressableWithSecondaryInteraction` component since selection is passed from it:
```js
executeSecondaryInteractionOnContextMenu(e) {
    const selection = SelectionParser.parseToMarkdown();
    e.stopPropagation();
    if (this.props.preventDefaultContentMenu) {
        e.preventDefault();
    }
    this.props.onSecondaryInteraction(e, selection);
}
```

Changes:
https://github.com/kursat/App/commit/84f4ba18f9face9bc792999c8a9bbd6253e6db54#diff-c352943cfbc37be8eb6e327927c6f50b353de8481ea7bc0bc1d40776513a301d
https://github.com/kursat/App/commit/84f4ba18f9face9bc792999c8a9bbd6253e6db54#diff-ec884d345ea3b9c8d2c1f65f808daa5b6a2f054e3c574aa724e13fd1de1fc717


## SelectionParser

Now this is a very basic draft that only parses messages with bold, italic, strike-through markdown. This will be developed to match all rules of `ExpensiMark`.

```js
import * as cheerio from 'cheerio';
import ExpensiMark from 'expensify-common/lib/ExpensiMark';

const getHTMLOfSelection = () => {
    if (window.getSelection) {
        const selection = window.getSelection();

        if (selection.rangeCount > 0) {
            const div = document.createElement('div');


            for (let i = 0; i < selection.rangeCount; i++) {
                const range = selection.getRangeAt(i);

                const clonedSelection = range.cloneContents();

                if (clonedSelection.textContent.length > 0) {
                    if (clonedSelection.childNodes.length > 0) {
                        Array.from(clonedSelection.childNodes).forEach((j) => {
                            if (j.textContent.trim().length > 0) {
                                div.appendChild(j);
                                div.appendChild(document.createElement('br'));
                            }
                        });
                    } else {
                        div.appendChild(clonedSelection);
                        div.appendChild(document.createElement('br'));
                    }
                }
            }

            // TODO: remove last br
            // TODO: dont remove inner line breaks

            return div.innerHTML;
        }

        return '';
    }

    return '';
};

const parseToMarkdown = () => {
    const selection = getHTMLOfSelection();

    const $ = cheerio.load(selection, null, false);

    const tags = ['strong', 'em', 'del'];

    tags.forEach((tag) => {
        $(`[data-testid="${tag}"]`).each((i, item) => {
            item.attribs = {};
            item.name = tag;
        });
    });

    const parser = new ExpensiMark();

    return parser.htmlToMarkdown($.html());
};

const SelectionParser = {
    parseToMarkdown,
};

export default SelectionParser;

```

### What it does:

1. Gets selection.
2. Iterates all ranges to parse from multiple messages.
3. Append all non-empty nodes to a div.
4. Loads contents of div to cherioo and replaces tags with attribute `data-testid`.
5. Parse clean HTML with `ExpensiMark` and return it.

### What needs to be done:

1. Implement all rules of `ExpensiMark`.
2. If we should, we can try to remove cherioo, and replace content with direct DOM manipulation.
3. Fix bugs (I'll add a video of both features and bugs)

Code: 
https://github.com/kursat/App/commit/84f4ba18f9face9bc792999c8a9bbd6253e6db54#diff-f131ebb969c388eeef6c2e8149c2b4a1aef340788fc743437d90f8e51fad9828

### Video:

https://user-images.githubusercontent.com/1279379/143324980-b09bfe76-5ac4-4bf6-b72d-29fb2b45c27e.mp4

**Comment by thienlnam** [comment]
Hey @kursat, thanks for the proposal - it looks good to me 🟢 

@dylanexpensify Could you please hire @kursat for the job?

**Comment by parasharrajat** [comment]
Wow, Cheerio is 300KB https://bundlephobia.com/package/cheerio@1.0.0-rc.10.

**Comment by thienlnam** [comment]
Great catch on keeping the bundle size small @parasharrajat, and yes if we could replace the use of cheerio with htmlparser2 in your proposal that would be great

**Comment by kursat** [comment]
@parasharrajat, nice suggestion, I was going to parse myself. Thank you.
> htmlparser2 is indirectly used by `react-native-render-html` so it won't increase bundle size. cc: @thienlnam

@thienlnam sure, I'll.

**Comment by kursat** [comment]
Hey @thienlnam, @dylanexpensify, I'am still not hired on upwork. Is there a problem/second thoughts with my proposal?

**Comment by thienlnam** [comment]
Hey @kursat, sorry for the delay.  No second thoughts about your proposal but just catching up after some time off.  I've pinged the project manager and we'll get you hired within the next day

**Comment by kursat** [comment]
Great @thienlnam, just wondered if I did something wrong on my proposal, no worries about delay. Thanks for your help.

**Comment by dylanexpensify** [comment]
Sending invite now!

**Comment by dylanexpensify** [comment]
@kursat offer sent! @parasharrajat mind applying also to the job so I can hire you for the reporting aspect?

**Comment by parasharrajat** [comment]
@dylanexpensify Applied

**Comment by kursat** [comment]
Hey I've started and almost finished coding, but I've came across to this issue that selecting text from multiple comment is not possible right now (It's still working for multiline comments, just not working for multi-comments). It was when I proposed, but with latest changes, comments are inserted to dom reversed. I reported this as bug, but I am not sure should we consider this a bug.

Please see current behavior below:

https://user-images.githubusercontent.com/1279379/145456055-90c14c8c-2a66-4958-82a7-15998f23abc1.mp4

What should I do next? Should I wait this to be fixed, or should I create my PR, ignoring this issue?

**Comment by parasharrajat** [comment]
https://github.com/Expensify/App/issues/5142#issuecomment-990123669

This is the biggest bounty issue on the repo and is being actively worked on. You can continue with your PR. Even if it is putting content in reverse order, that is fine. just that formatting should be preserved.

**Comment by kursat** [comment]
Hey great, I've deleted my message from slack. Proceeding to work on it. Thanks for quick reply and great info @parasharrajat.

**Comment by dylanexpensify** [comment]
@parasharrajat sent offer!

**Comment by thienlnam** [comment]
PR in review here: https://github.com/Expensify/App/pull/6732

Changes are looking good but we're working through a package-lock.json diff

**Comment by thienlnam** [comment]
Just merged the PR - now waiting on the regression period (7 days)

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.24-19 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/6732
- https://github.com/Expensify/App/pull/6732


If no regressions arise, payment will be issued on 2022-01-11. :confetti_ball:

**Comment by mallenexpensify** [comment]
@thienlnam sanity check... before the issue, when you pasted a url into NewDot it didn't auto-hyperlink, you'd have to include markdown.   Now, after this PR/issue, when you paste a link in, it's auto-hyperlinked? and.. it doesn't show the markdown it used to ?

**Comment by thienlnam** [comment]
Chatted with @mallenexpensify 1:1

It looks like the fix is not working correctly - it does hyperlinks well but for other markdown options it does not work as expected

Sent text:
<img width="395" alt="Screen Shot 2022-01-04 at 11 00 34 AM" src="https://user-images.githubusercontent.com/30609178/148110025-02913072-e59c-4957-89a9-ab039a6957d7.png">

When copied and pasted:
<img width="334" alt="Screen Shot 2022-01-04 at 11 00 56 AM" src="https://user-images.githubusercontent.com/30609178/148110054-b8a29e01-db2e-485d-9e36-dbb6c46585ee.png">

cc @kursat

**Comment by MelvinBot** [comment]
📣 @kursat You have been assigned to this job by @thienlnam!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by kursat** [comment]
Hey @thienlnam, @mallenexpensify,

I am not able to reproduce issue on rev. 580c9dbb or on https://staging.new.expensify.com. Tried both on Firefox and Chrome on Ubuntu 20.04.3 LTS. Also checked context menus both Android and Web versions.

Can you provide details about the issue? What was the copying method? What's your environment?

**Comment by thienlnam** [comment]
We're testing this on production since the PRs have been deployed there.

Currently testing on Chrome on a Mac V11.6

Tested with copying using the context menu
<img width="60" alt="Screen Shot 2022-01-05 at 3 03 33 PM" src="https://user-images.githubusercontent.com/30609178/148302153-98bee104-a47c-4d8f-88dd-4de6e59b89cf.png">

and using CMD + V to paste

as well as copying using CMD + C and using CMD + V to paste

**Comment by kursat** [comment]
Its working on following env:

Chrome: Version 97.0.4692.71 (Official Build) (x86_64)
MacOS: 10.15.7 (19H1615)

https://user-images.githubusercontent.com/1279379/148338723-df6d14b2-2f35-46c6-94ae-7d22be777b0f.mp4

> Tested with copying using the context menu <img alt="Screen Shot 2022-01-05 at 3 03 33 PM" width="60" src="https://user-images.githubusercontent.com/30609178/148302153-98bee104-a47c-4d8f-88dd-4de6e59b89cf.png">

Also tried with hover menu which I noticed after shooting the video.

Currently this is what i got but I'll try to find a way to test on newer versions of mac.

**Comment by parasharrajat** [comment]
It should be platform agnostic. Maybe we are missing some step here.

**Comment by thienlnam** [comment]
Huh, I retested this morning and it is working as expected - maybe the deploy didn't uninstall on prod until recently.  Either way, all good here - thanks for confirming!

**Comment by dylanexpensify** [comment]
Payment sent for both @parasharrajat and @kursat! Contracts ended, job posting closed

**Comment by dylanexpensify** [comment]
@parasharrajat created new job for you to apply to to get paid for reporting via agency account! https://www.upwork.com/jobs/~01c306c74dd75512bf

**Comment by dylanexpensify** [comment]
offer sent

**Comment by dylanexpensify** [comment]
payment sent, job closed, contract ended fr Rajat
